dpgen/generator/model_devi.py

```diff
@@ -0,0 +1,191 @@
+        conf_name: str
+            file name of initial system
+        models: list[str]
+            list of models
+        """
+        raise NotImplementedError("Not implemented")
+
+    @abstractmethod
+    def get_running_parameters(self, work_path: str) -> Tuple[str, List[str]]:
+        """Get running parameters.
+
+        Parameters
+        ----------
+        work_path: str
+            the work path
+
+        Returns
+        -------
+        command: str
+            the running command
+        forward_files: list of str
+            forward files
+        backward_files: list of str
+            backward files
+        common_files: list of str
+            common files
+        """
+        raise NotImplementedError("Not implemented")
+
+
+    @abstractmethod
+    def extract_trajectory(self, directory) -> 'Trajectory':
+        """Obtain a Trajectory object from a directory.
+        
+        Returns
+        -------
+        trajectory: Trajectory
+            the Trajectory object
+        """
+        return Trajectory(directory)
+
+
+class Trajectory(ABC):
+    """A trajectory containing lots of frames or clusters.
+    
+    Parameters
+    ----------
+    engien: ModelDeviEngien
+        Engien
+    directory: str
+        The directory of a simulation
+    """
+    def __init__(self, engien: ModelDeviEngien, directory: str) -> None:
+        self.engien = engien
+        self.directory = directory
+        self.status = 0
+    
+    @abstractmethod
+    def get_model_deviations(self) -> np.ndarray:
+        """Get model deviations of all frames or clusters.
+
+        Returns
+        -------
+        model_deviations: np.ndarray[float], 1D or 2D
+            Model deviation of all frames or all clusters. 
+            - frame: 1D, the maximum model deviation of all atoms
+            - clusters: 2D, the atomic model deviation of centeral atom
+        
+        Notes
+        -----
+        The index of model deviation should be as the same as frames.
+        """
+        # for performance, returns a numpy array instead of a list
+        raise NotImplementedError("Not implemented")
+    
+    @abstractmethod
+    def get_frame(self, idx: Tuple[int]) -> "Frame":
+        """Get a frame from idx.
+
+        Parameters
+        ----------
+        idx: tuple[int]
+            Index.
+        
+        Returns
+        -------
+        frames: Frame
+            Frames.
+        """
+        return Frame(self, idx)
+    
+    def open_trajectory(self):
+        """Open the trajectory.
+        
+        Notes
+        -----
+        For example, open the file stream, initialize the parameters,
+        and others.
+        """
+        self.status = 1
+
+    def close_trajectory(self):
+        """Close the trajectory. Call this method when all frames have
+        been read.
+        """
+        self.status = 0
+    
+    def __lt__(self, other: "Trajectory") -> bool:
```

Comment:
Is it a good behaviour to compare two strings？

Reply:
The idea of comparing is let frames with the same trajectory open together, considering some types of trajectories have only one file to read all frames. So, the way to sort `Trajectory` is not important. It's just used to sort `Frame`:
https://github.com/deepmodeling/dpgen/blob/11a77abba134d710ef5881c6a19da6f3e0972a2d/dpgen/generator/model_devi.py#L187-L191